doc/source/development/rfc/rfc92_wkb_only_geometries.rst

```diff
@@ -0,0 +1,228 @@
+  honour RequestWKBOnlyGeometries(true)
+* OLCWriteWKBGeometries: a layer must return TRUE for it if its CreateFeature()
+  and SetFeature() implementations support being passed OGRWKBOnlyGeometry
+  instances.
+
+ogr2ogr is modified to call RequestWKBOnlyGeometries(true) on the source layer:
+
+* if the source layer advertises OLCReadWKBGeometries
+* if the target layer advertises OLCWriteWKBGeometries
+* if no command line switch requires a "materialized" geometry.
+
+So basically this is for requests like
+"ogr2ogr out.gpkg in.gpkg [layer or SQL request] [attribute filter] [spatial filter]"
+
+C API
+-----
+
+OGR_L_RequestWKBOnlyGeometries() is added.
+
+Backward compatibility
+----------------------
+
+No issue. Only API additions
+
+Benchmarking
+------------
+
+bench_ogr_c_api
+~~~~~~~~~~~~~~~
+
+The `bench_ogr_c_api <https://github.com/OSGeo/gdal/blob/master/perftests/bench_ogr_c_api.cpp>`_
+benchmark utility which uses the C API to iterate over features and get their WKB
+representation is enhanced with a `-wkb_only_geometry` switch to call
+OGR_L_RequestWKBOnlyGeometries().
+
+On a 1.6 GB GeoPackage (nz-building-outlines.gpkg) made of 3.2 million features
+with simple polygons (typically quadrilaterals, building footprints) and 13
+attributes:
+
+- bench_ogr_c_api runs in 6.4 s
+- bench_ogr_c_api -wkb_only_geometry runs in 5.0 s
+
+==> 22% faster
+
+Other synthetic benchmarks show that the maximum speed-up is about 30% on a
+dataset with 10 millions polygonal features of 10 points each.
+
+Conversely, the gain is much more modest, or close to null, with just a few
+thousands of features that hold larger geometries (several thousands of points
+each).
+
+The gain is more in saving instantiation of OGRPolygon and OGRLinearRing
+objects that in the size of their coordinate set.
+
+ogr2ogr
+~~~~~~~
+
+::
+
+    ogr2ogr /vsimem/out.gpkg nz-building-outlines.gpkg -lco spatial_index=no
+
+
+runs in 15.8 second in WKBOnlyGeometry mode vs 19.1 second without it (master),
+hence a 17% speed-up.
+
+With spatial index creation enabled (multi-threaded), the wall clock time
+difference is within measurement noise. And for singe threaded creation, the
+WKBOnlyGeometry mode is 5% faster.
+
+Discussion
+----------
+
+Is it a good idea... ?
+
+The design of OGRWKBOnlyGeometry is admittedly a bit clunky, or at least at odds
+with other OGRGeometry subclasses, but nothing more elegant, concise, performant
+and that doesn't change the whole OGRGeometry API and driver implementations
+comes to mind.
+
+The scope is limited to a few drivers: GeoPackage, PostGIS (but the current
+throughput of the driver is probably not sufficient for OGRGeometry overhead to
+be noticeable), what else?
+
+Should methods of OGRWKBOnlyGeometry that cannot work without materialization
+of the real geometry return an error like done currently, or do the materialization
+on-the-fly when needed ? The motivation for erroring out is to avoid silent
```

Comment:
At worst these are going to be ~same cost as current code, so I don't see the cost of materializing when needed as a blocker. I'd lean toward materialization on demand make it simpler/safer for users to always just ask for WKB only and trust GDAL to do the right thing transparently.

Reply:
> I'd lean toward materialization on demand 

That isn't as simple as that. For example as commented in the prototype of the class, getGeometryType() cannot safely be implemented as Materialize()->getGeometryType(), because otherwise users would wrongly cast a OGRWKBOnlyGeometry* to a OGRPoint*, OGRLineString*, etc

Reply:
I guess it's not a trivial change, but we could introduce a `ogrgeometry_cast` implementation which has special handling for OGRWKBOnlyGeometry and correctly returns the cast materialized geometry.

Reply:
The base OGRGeometry class has *non virtual* toPoint(), toPolygon(), etc. cast operators. So if there were made virtual (but this would have a small run-time cost), potentially OGRWKBOnlyGeometry could reimplement them (that would also involve that OGRWKBOnlyGeometry stores the materialized geometry, when creating it). But that assumes that all code that cast OGRGeometry* to a sub-class uses those operators, which is very likely not true as they have been introduced relatively recently (I remember having made a pass to use those operators, but I'm pretty sure it isn't exhaustive, as there's no way to spot all code paths which do those casts. And even if we could do it in the GDAL code base, that would also impact external code using GDAL C++ API).